src/core/ddsc/tests/psmxif.c

```diff
@@ -0,0 +1,441 @@
+// Copyright(c) 2020 to 2023 ZettaScale Technology and others
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+// v. 1.0 which is available at
+// http://www.eclipse.org/org/documents/edl-v10.php.
+//
+// SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+
+#include <assert.h>
+#include <limits.h>
+
+#include "dds/ddsrt/mh3.h"
+#include "dds/ddsrt/md5.h"
+#include "dds/ddsrt/io.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/string.h"
+#include "dds/ddsrt/bswap.h"
+#include "dds/ddsrt/environ.h"
+#include "dds/ddsrt/static_assert.h"
+
+#include "dds/dds.h"
+#include "dds/ddsi/ddsi_entity_index.h"
+#include "ddsi__addrset.h"
+#include "ddsi__entity.h"
+#include "ddsi__xevent.h"
+#include "dds__entity.h"
+#include "dds__serdata_default.h"
+
+#include "config_env.h"
+#include "test_common.h"
+#include "psmx_dummy_public.h"
+#include "Array100.h"
+#include "DynamicData.h"
+#include "PsmxDataModels.h"
+
+static void free_strings(uint32_t len, char** strings)
+{
+    if (len != 0 && strings != NULL) {
+        for (uint32_t i = 0; i < len; i++) {
+            dds_free(strings[i]);
+        }
+    }
+    dds_free(strings);
+}
+
+/**
+ * @brief Create a participant object, providing a configuration to choose the psmx interface.
+ * 
+ * @param[in] int_dom the domain id
+ * @param[in] cdds_psmx_name the name of the psmx interface to use
+ * @param[in] locator an array of 16 bytes, NULL if not used
+ * @return dds_entity_t the participant
+ */
+static char* create_config(dds_domainid_t int_dom, const char* cdds_psmx_name, const uint8_t* locator)
+{
+  char *configstr;
+  char locator_str[74];
+  if ( locator == NULL ) {
+    locator_str[0] = 0;
+  } else {
+    const uint8_t *l = locator;
+    snprintf(
+      locator_str,
+      sizeof(locator_str),
+      "LOCATOR=%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x;",
+      l[0], l[1], l[2], l[3], l[4], l[5], l[6], l[7], l[8], l[9], l[10], l[11], l[12], l[13], l[14], l[15]
+    );
+  }
+  ddsrt_asprintf (&configstr, "\
+${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}\
+<General>\
+  <AllowMulticast>spdp</AllowMulticast>\
+  <Interfaces>\
+    <PubSubMessageExchange name=\"%s\" library=\"psmx_%s\" priority=\"1000000\" config=\"%sSERVICE_NAME=psmx%d;KEYED_TOPICS=true;\" />\
+  </Interfaces>\
+</General>\
+<Discovery>\
+  <Tag>${CYCLONEDDS_PID}</Tag>\
+  <ExternalDomainId>0</ExternalDomainId>\
```

Comment:
This just runs a single instance of Cyclone, no need for the tricks that allow you to pretend there's a network in between when everything is in a single process.